### ISSUE TITLE
Add option to `require-default-props` to forbid default props for required props

### ISSUE THREAD

**Issue by sharmilajesupaul** [yannickcr/eslint-plugin-react#1524]
The [`require-default-props`](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/require-default-props.md) rule allows required props to also have a default prop. It would be nice to add an option to warn against this behavior. 

Option:
`forbidDefaultForRequired` (`boolean`) - Setting this to `true` will forbid required props from also having a default prop declaration. Default `false`.

```js
schema: [{
 type: 'object',
      properties: {
        forbidDefaultForRequired: {
          type: 'boolean'
        }
      },
      additionalProperties: false,
}]
```
The following patterns are considered warnings:
```js
function MyStatelessComponent({ foo, bar }) {
  return <div>{foo}{bar}</div>;
}

MyStatelessComponent.propTypes = {
  foo: PropTypes.string.isRequired,
};

MyStatelessComponent.defaultProps = {
  foo: 'something',
};
```

```js
class Foo extends React.Component {
  render() {
    return (
      <h1>Hello, {this.props.foo}</h1>
    );
  }

  static propTypes = {
    foo: PropTypes.string,
  };

  static defaultProps = {
    foo: "foo",
  };
}
```

The following patterns are **not** considered warnings:
```js
function MyStatelessComponent({ foo, bar }) {
  return <div>{foo}{bar}</div>;
}

MyStatelessComponent.propTypes = {
  foo: PropTypes.string,
};

MyStatelessComponent.defaultProps = {
  foo: 'something',
};
```
```js
class Foo extends React.Component {
  render() {
    return (
      <h1>Hello, {this.props.foo}</h1>
    );
  }

  static propTypes = {
    foo: PropTypes.string,
  };

  static defaultProps = {
    foo: "foo",
  };
}
```

**Comment by ljharb** [comment]
Sound great to me - as long as it does a "safe" thing when it encounters things that aren't statically analyzeable, like imported shapes, spread props, etc. It'd also be super fancy if spread props worked when the object literal it referenced was defined in the same file.

**Comment by sharmilajesupaul** [comment]
+1 on warning on spread prop violations when the referenced object literal in the same file

Is the "safe" thing generally not warning on propTypes that aren't explicitly declared in the same file?

**Comment by ljharb** [comment]
Yes, exactly - basically, what's safe is not creating false positives.
